### PR TITLE
Allow parsing sub-sat denominations with decimal points 

### PR DESCRIPTION
**Pull request by Kixunil**
Numbers with only zeros after decimal points are valid if they are also
multiples of `10^precision` (e.g. 1000 for msats). These were
artificially disallowed as "too precise" which was at least misleading.

This change allows parsing such numbers.

And yes, I know this is not perfectly efficient (unless the compiler figures out some magic opts) but so isn't the rest of the code. TBH this parsing code drives me crazy and I'd love to rewrite it to be more efficient and readable.

### CODE REVIEW

**Comment by apoelstra** [comment]
I'm skeptical that we should allow too-precise values just because the extra precision happens to specify something we're able to represent exactly. It seems like a special case that users would need to memorize. Especially if they changed the data and suddenly got a "too precise" error even though their precision hadn't changed at all.

In general I've been willing to ACK basically anything related to this code because I don't care about trying to parse amounts from strings, and don't plan to ever care, and I can see the argument for this, so I'm happy to ACK it. But I also don't necessarily think it's the right idea.

**Comment by Kixunil** [comment]
My idea is to be consistent with `f64` which always allows decimal point. The case of people changing data and suddenly becoming an error is already a problem e.g. we accept `1000 msat` so changing it to `1001 msat` will trigger the error you describe.

**Comment by apoelstra** [comment]
Well, `f64` will just silently mangle data rather than producing precision errors under any circumstances.

But ok, if these are the existing semantics of `msat` etc., then I agree that this PR is consistent with that. So I rescind my skepticim.

**Review by tcharding — APPROVED** [review]
ACK b48065fc723f9128868e306a2316691dd7e469c2

**Comment by apoelstra** [comment]
1176ae5f58ef6d81d23d5dc6a66ebc1f5f1154ca does not compile.
```
error[E0425]: cannot find value `msat` in this scope
    --> bitcoin/src/amount.rs:1657:37
     |
1657 |         assert_eq!(p("999.0000000", msat), Err(E::TooPrecise));
     |                                     ^^^^ help: a local variable with a similar name exists: `sat`

```

File: bitcoin/src/amount.rs [hunk 1]
```diff
@@ -199,7 +199,13 @@ impl std::error::Error for ParseAmountError {
 }
 
 fn is_too_precise(s: &str, precision: usize) -> bool {
-    s.contains('.') || precision >= s.len() || s.chars().rev().take(precision).any(|d| d != '0')
+    match s.find('.') {
+        Some(pos) =>
+            s[(pos + 1)..].chars().any(|d| d != '0')
```
Comment (stevenroose) [inline bitcoin/src/amount.rs:204]: nit: I think `s.chars().skip(pos + 1)` is more readable here

**Review by stevenroose — COMMENTED** [review]
Just 2 nits, feel free to ignore if you disagree. At least I guess the suggested changes make the code more alike the surrounding code (which happens to be mostly mine..)

**Comment by stevenroose** [comment]
concept ACK, though tbh I agree with Andrew. In principle it would be better to no longer allow 1000 msat instead of allowing more off things.

What is the motivation for this change? Did you bump into this restruction personally?

File: bitcoin/src/amount.rs [hunk 2]
```diff
@@ -199,7 +199,13 @@ impl std::error::Error for ParseAmountError {
...
+                || precision >= pos
+                || s[..pos].chars().rev().take(precision).any(|d| d != '0'),
```
[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/src/amount.rs:206]: Same problem as above.

**Comment by Kixunil** [comment]
@stevenroose the main motivation is consistency. Anything that is a correctly formatted number should be accepted if it's representable. Note that I also find the limit on maximum length kinda weird but I guess it's DoS protection and the error message is reasonable. I guess there was a reason to support msats in the first place.

**Pull request merged**